### PR TITLE
[WEB-1344-46] Curve Registry Override

### PR DESCRIPTION
**Pull request by pmdaly**
Curve doesn't always update their registry to reflect the most recent LP->Pool address mappings.   This PR creates an override that allows Yearn to add additional mappings.  Although 15 files are changed, many of these are b/c of linting updates; the files of concern are primarily

[contracts/Registries/Overrides.sol](https://github.com/yearn/yearn-lens/pull/46/files#diff-b64fed0b791ca0879473fcde007d1d2e8ac3d1819de21a9e873ed8bc978a91a0)
[tests/registries/test_curve_override.py](https://github.com/yearn/yearn-lens/pull/46/files#diff-58aa29f4cd7b9c5de8c9da99f3f400b85c3c3979a3564d434d5f39df5fbebce2)

### CODE REVIEW

**Review by RM-0311 — DISMISSED** [review]

File: contracts/Registries/Overrides.sol [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+interface ICurveAddressesProvider {
+    function max_id() external view returns (uint256);
+
+    function get_address(uint256) external view returns (address);
+}
+
+contract CurveRegistryOverrides is Ownable {
+    ICurveAddressesProvider public curveAddressesProvider;
+    ICurveRegistry public curveRegistry;
```
Comment (jstashh) [inline contracts/Registries/Overrides.sol:23]: not sure if you already know this, but you can get the registry from the address provider

File: contracts/Registries/Overrides.sol [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    function addCurveRegistry(address _curveRegistryAddress) public onlyOwner {
+        require(
+            _curveRegistryAddress != address(0),
+            "Missing Curve Registry Address"
+        );
+        curveRegistry = ICurveRegistry(_curveRegistryAddress);
+    }
+
+    /// @notice Returns both override and curve registry pools as list
+    function curveRegistriesList() public view returns (address[] memory) {
```
Comment (jstashh) [inline contracts/Registries/Overrides.sol:55]: Do we need this function? I don't think a pool is the same thing a registry as well

File: contracts/Registries/Overrides.sol [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        if (pool != address(0)) {
+            return pool;
+        }
+        pool = executeSelectorOnCurveRegistries(_lpAddress);
+        return pool;
+    }
+
+    /// @notice Cycle through all curve registries and try to staticcall each one using manual selector
+    /// @dev return type is inconsistent across curve registres
+    function executeSelectorOnCurveRegistries(address _lpAddress)
```
Comment (jstashh) [inline contracts/Registries/Overrides.sol:91]: Not sure about the naming of this function, it sort of implies you should be able to pass in a selector but we always use `get_pool_from_lp_token `

File: contracts/Registries/Overrides.sol [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        view
+        returns (address)
+    {
+        uint256 numRegistries = curveAddressesProvider.max_id() + 1;
+        address curveRegistryAddress;
+        address pool;
+        bool success;
+        bytes memory data;
+        for (uint256 i; i < numRegistries; i++) {
+            curveRegistryAddress = curveAddressesProvider.get_address(i);
```
Comment (jstashh) [inline contracts/Registries/Overrides.sol:102]: Don't think every address from the address provider is necessarily a registry, it's only addresses at positions 0 and 5 from memory

**Review by jstashh — COMMENTED** [review]
Thanks for working on this, it'll be helpful to integrate it. 

I think it's unnecessary to loop through all the addresses that the address provider has, since not all of them are registries, maybe instead we should not use the address provider, and just have an array of registries we can manually add to/remove from, then we can just loop through those. What do you think? I think there are only two registries at the moment

File: contracts/Registries/Overrides.sol (hunk 1 above)
Comment (pmdaly) [reply contracts/Registries/Overrides.sol]: Updated to use addresses provider to retrieve registries.

File: contracts/Registries/Overrides.sol (hunk 2 above)
Comment (pmdaly) [reply contracts/Registries/Overrides.sol:33]: You're right.  Updated to return the curve registries as a list.  Function was an original design requirement, will discuss with @x48115

File: contracts/Registries/Overrides.sol (hunk 3 above)
Comment (pmdaly) [reply contracts/Registries/Overrides.sol]: Naming needs to be more explicit, got it.  Function ended up being removed preferring querying the exact registries instead of just trying the call.

File: contracts/Registries/Overrides.sol (hunk 4 above)
Comment (pmdaly) [reply contracts/Registries/Overrides.sol]: Updated to query the 0th and 5th.

File: contracts/Registries/Overrides.sol [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        require(
+            _curveAddressesProviderAddress != address(0),
+            "Missing Curve Addresses Provider address"
+        );
+        curveAddressesProvider = ICurveAddressesProvider(
+            _curveAddressesProviderAddress
+        );
+    }
+
+    /// @notice Returns all curve registries
```
Comment (jstashh) [inline contracts/Registries/Overrides.sol:31]: Don't think all addresses provided by the address provider are necessarily registries, not really an issue but just don't want it to be misleading

File: contracts/Registries/Overrides.sol [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+    /// @notice Search through pool registry overrides and curve registries for a LP Pool
+    /// @dev the 0th and 5th curve registry addresses contain LP -> pool mappings
+    function poolByLp(address _lpAddress) public view returns (address) {
+        address pool = poolByLpOverride[_lpAddress];
+        if (pool != address(0)) {
+            return pool;
+        }
+        // check 1st pool from registry provider
+        pool = getCurveRegistry(0).get_pool_from_lp_token(_lpAddress);
```
Comment (jstashh) [inline contracts/Registries/Overrides.sol:68]: Just wondering what will happen if a new registry is added, in a new position, we would have to modify this contract and deploy it again. It might be better if we had an array of registries we could add/remove from, to manually maintain, so it's flexible if a new one is added or if one is removed. In which case we wouldn't need to read anything from the address provider in this contract

Comment (pmdaly) [reply contracts/Registries/Overrides.sol]: updated so the contract owner supplies the registries and can update them as well

File: contracts/Registries/Overrides.sol (hunk 5 above)
Comment (pmdaly) [reply contracts/Registries/Overrides.sol]: updated function to return all curve registry addresses that are currently set

File: contracts/Libraries/String.sol [hunk 7]
```diff
@@ -1,5 +1,5 @@
 // SPDX-License-Identifier: MIT
-pragma solidity 0.8.11;
+pragma solidity ^0.8.11;
```
Comment (jstashh) [inline contracts/Libraries/String.sol:2]: is this change needed? I think we generally pin to exact versions

**Review by jstashh — COMMENTED** [review]
LGTM! just one question but otherwise we'll be able to merge 🙂

**Review by jstashh — APPROVED** [review]
